### ISSUE TITLE
Support Apple Silicon (macOS arm64) and Xcode 12

### ISSUE THREAD

**Issue by gaige** [microsoft/plcrashreporter#115]
When compiling under Xcode 12 with macos arm64 target, an error is raised in plcrash_async_cfe_entry_return_address_register due to what appears to be an errant boolean expression using a constant.

Docs indicate that the entry types  `PLCRASH_ASYNC_CFE_ENTRY_TYPE_FRAMELESS_IMMD` and `PLCRASH_ASYNC_CFE_ENTRY_TYPE_FRAMELESS_INDIRECT` should be supported, but the boolean expression appears to || in the  `PLCRASH_ASYNC_CFE_ENTRY_TYPE_FRAMELESS_INDIRECT`.

It looks like the fix is to add (entry->type ==) around the `PLCRASH_ASYNC_CFE_ENTRY_TYPE_FRAMELESS_INDIRECT`, which allows it to compile and appears consistent with the documentation of the function.

**Comment by MatkovIvan** [comment]
Hi @gaige,

Thanks for the report. PLCrashReporter doesn't support Apple Silicon and Xcode 12 yet. In our case it is not "just rebuild" for another architecture. I'm marking this as a feature request

**Comment by gaige** [comment]
I understand, and while that's true, I think this warrants some further investigation as the expression in question:

````
    PLCF_ASSERT(entry->return_address_register == PLCRASH_REG_INVALID || (entry->type == PLCRASH_ASYNC_CFE_ENTRY_TYPE_FRAMELESS_IMMD || PLCRASH_ASYNC_CFE_ENTRY_TYPE_FRAMELESS_INDIRECT));
````

Is always true, since `PLCRASH_ASYNC_CFE_ENTRY_TYPE_FRAMELESS_INDIRECT` is a non-zero constant.  The reason this got caught by Xcode 12 is because the `bool` type in Apple Silicon is more stringent, however this doesn't change the liklihood that an assert which is always true is not performing any function.

**Comment by jakob** [comment]
This specific issue should be fixed by my PR #111 which was just merged.

**Comment by duncanwilcox** [comment]
We need crash reports. We need to support Apple Silicon. Is plcrashreporter realistically going to be updated to support Apple Silicon by the end of the year?

**Comment by adib** [comment]
What's the core difference between Apple Silicon on the Mac vs the crash reporting functionality that is already shipped for iOS such that this became a _feature request_ ?

**Comment by jakob** [comment]
We did some superficial testing and it seems that PLCrashReporter works just fine on Apple Silicon if you compile it yourself.

I'm assuming that Microsoft doesn't want to officially claim that they support Apple Silicon without proper testing etc. There are probably still some bugs to fix. For example, in our initial tests crash reports did not show the correct computer model when running in Rosetta. (Which might be an issue with Rosetta, since the same error occurs in other apps)

**Comment by jakob** [comment]
(Quick Follow Up: The computer model bug in Rosetta has been fixed in a recent beta of Big Sur)

**Comment by MatkovIvan** [comment]
1.8.0 with Apple Silicon support has been released. Please give it a try and let us know if you have any issues. Thanks

**Comment by duncanwilcox** [comment]
Thanks @MatkovIvan. I haven't looked at it yet, but just a warning for everybody planning to use the binary distribution, in case it's built with a beta Xcode (say 12.2b3) it will be rejected by Apple.

**Comment by triplef** [comment]
Thanks for the heads up @duncanwilcox! Is that the case even for frameworks or libraries embedded in an app, i.e. would Apple reject even if the app itself is built with an XCode GM release?

At least for static framework and libraries I'd be surprised if there is even a way to detect which Xcode version was used for compilation, although for dynamic frameworks I think it's right in the Info.plist.

**Comment by duncanwilcox** [comment]
@triplef the rejection is generally due to either beta macOS or beta Xcode. Both need to be at least GM. I'm not really sure about static libraries but you might be right. We're rebuilding on 10.15.7 when Xcode GM is out (guessing Nov 10), just in case.

**Comment by MatkovIvan** [comment]
@duncanwilcox I just published [1.8.1](https://github.com/microsoft/plcrashreporter/releases/tag/1.8.1) which I built with [Xcode 12.2 Release Candidate](https://developer.apple.com/news/releases/?id=11052020h) to [be able to](https://developer.apple.com/news/releases/?id=11052020i) submit the applications that use the framework as a binary to the App Store.
